chemprop/cli/train.py

```diff
@@ -542,170 +531,220 @@ def main(args):
+                criterion=criterion,
+                n_classes=args.multiclass_num_classes,
+                spectral_activation=args.spectral_activation,
+            )
+
+            if isinstance(predictor_ffn, RegressionFFN):
+                scaler = train_dset.normalize_targets()
+                val_dset.normalize_targets(scaler)
+                logger.info(f"Train data: loc = {scaler.mean_}, scale = {scaler.scale_}")
+            else:
+                scaler = None
+
+            train_loader = data.MolGraphDataLoader(train_dset, args.batch_size, args.num_workers)
+            val_loader = data.MolGraphDataLoader(
+                val_dset, args.batch_size, args.num_workers, shuffle=False
+            )
+
+            if multicomponent:
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+            else:
+                if len(test_data) > 0:
+                    test_dset = make_dataset(test_data[fold_idx], args.rxn_mode)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+
+            mpnn_cls = MulticomponentMPNN if multicomponent else MPNN
+            model = mpnn_cls(
+                mp_block,
+                agg,
+                predictor_ffn,
+                True,
+                None,
+                args.task_weights,
+                args.warmup_epochs,
+                args.init_lr,
+                args.max_lr,
+                args.final_lr,
+            )
+            logger.info(model)
+
+            monitor_mode = "min" if model.metrics[0].minimize else "max"
+            logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+
+            if no_cv:
+                output_dir = args.output_dir
+            else:
+                output_dir = args.output_dir / f"fold_{fold_idx}"
+
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
+                "{epoch}-{val_loss:.2f}",
+                "val_loss",
+                mode=monitor_mode,
+                save_last=True,
+            )
+
+            early_stopping = EarlyStopping("val_loss", patience=5, mode=monitor_mode)
+
+            trainer = pl.Trainer(
+                logger=tb_logger,
+                enable_progress_bar=True,
+                accelerator="auto",
+                devices=args.n_gpu if torch.cuda.is_available() else 1,
```

Comment:
There's a bit of difficulty in this - lightning has a lot of ways to use `devices`, see: https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#:~:text=The%20table%20below%20lists%20examples%20of%20possible%20input%20formats%20and%20how%20they%20are%20interpreted%20by%20Lightning.

We should follow up about this offline. The way we pass this arg should probably change if we want to expose all the functionality of `devices`.

Reply:
Agreed. IMO we should have an `int` arg `--devices` that's just the _number_ of devices to train on. If a user wants to limit the GPU set, they can just run it like so:
```
$ CUDA_VISIBLE_DEVICES=0,2 chemprop train --devices -1
```
or we could also use the `find_usable_cuda_devices` method so that a user doesn't even need to bother with limiting GPU visibility